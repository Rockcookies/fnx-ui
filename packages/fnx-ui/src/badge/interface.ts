import { ReactNode, HTMLAttributes } from 'react';

export type BadgeProps = BadgeComponentProps &
	Omit<HTMLAttributes<HTMLSpanElement>, 'color'>;

export interface BadgeComponentProps {
	count?: ReactNode;
	showZero?: boolean;
	dot?: boolean;
	max?: string | number;
	color?: string;
	offset?: [number | string, number | string];
}
