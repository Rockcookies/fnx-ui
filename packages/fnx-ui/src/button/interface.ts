import { ButtonHTMLAttributes, ReactNode } from 'react';

export type ButtonType =
	| 'default'
	| 'primary'
	| 'success'
	| 'warning'
	| 'danger';

export type ButtonSize = 'lg' | 'md' | 'sm' | 'xs';

export type ButtonShape = 'rect' | 'radius' | 'round';

export interface ButtonComponentProps {
	type?: ButtonType;
	size?: ButtonSize;
	icon?: ReactNode;
	iconPosition?: 'left' | 'right';
	loading?: boolean;
	loadingIcon?: ReactNode;
	loadingChildren?: ReactNode;
	disabled?: boolean;
	hairline?: boolean;
	plain?: boolean;
	shape?: ButtonShape;
	block?: boolean;
	color?: string;
	htmlType?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
}

export type ButtonProps = ButtonComponentProps &
	Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'type'>;
