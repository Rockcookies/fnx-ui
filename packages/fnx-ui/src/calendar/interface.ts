import { HTMLAttributes, ReactNode } from 'react';
import CalendarDay from './CalendarDay';
import CalendarMulti from './CalendarMulti';
import CalendarRange from './CalendarRange';

export type CalendarValue = Date | string | number;

export type CalendarDayType =
	| 'normal'
	| 'start'
	| 'start-end'
	| 'middle'
	| 'end'
	| 'selected'
	| 'multiple-middle'
	| 'multiple-selected'
	| 'disabled';

export interface CalendarDayComponentProps {
	color?: string;
	dayHeight?: string;
	date: Date;
	content?: ReactNode;
	type: CalendarDayType;
	topInfo?: ReactNode;
	bottomInfo?: ReactNode;
}

export type CalendarDayProps = CalendarDayComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'content'>;

export interface CalendarSlots {
	day?: (day: CalendarDayComponentProps) => ReactNode;
	subTitle?: boolean | ((date: Date) => ReactNode);
	monthMark?: boolean | ((date: Date) => ReactNode);
	monthTitle?: boolean | ((date: Date) => ReactNode);
	footer?: ReactNode;
}

export interface CalendarBaseProps {
	title?: ReactNode;
	color?: string;
	minDate?: Date | string | number;
	maxDate?: Date | string | number;
	dayHeight?: number | string;
	readonly?: boolean;
	closeIcon?: ReactNode;
	firstDayOfWeek?: number;
	slots?: CalendarSlots;
	confirmText?: ReactNode;
	confirmDisabledText?: ReactNode;
}

export type CalendarElementProps = Omit<
	HTMLAttributes<HTMLDivElement>,
	'defaultValue' | 'onChange' | 'title'
>;

export interface CalendarContainerRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date[] | undefined;
}

export interface CalendarRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date | undefined;
}

export interface CalendarComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue;
	onChange?: (value: Date) => void;
	onConfirm?: (value?: Date) => void;
	onCancel?: () => void;
}

export type CalendarProps = CalendarComponentProps & CalendarElementProps;

export type CalendarRangeRef = CalendarContainerRef;

export interface CalendarRangeComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	allowSameDay?: boolean;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarRangeProps = CalendarRangeComponentProps &
	CalendarElementProps;

export type CalendarMultiRef = CalendarRangeRef;

export interface CalendarMultiComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarMultiProps = CalendarMultiComponentProps &
	CalendarElementProps;

export interface CalendarExportExtra {
	Range: typeof CalendarRange;
	Multi: typeof CalendarMulti;
	Day: typeof CalendarDay;
}
