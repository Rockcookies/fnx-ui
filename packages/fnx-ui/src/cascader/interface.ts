import { HTMLAttributes, ReactNode } from 'react';

export type CascaderValue = string | number;

export interface CascaderOption {
	label?: ReactNode;
	value?: CascaderOption;
	loading?: boolean;
	disabled?: boolean;
	children?: CascaderOption[];
	// for custom filed names
	[key: string]: any;
}

export interface CascaderDataNames {
	label?: string | string[];
	value?: string | string[];
	loading?: string | string[];
	disabled?: string | string[];
	children?: string | string[];
}

export interface CascaderComponentProps<T = CascaderOption> {
	defaultValue?: CascaderValue[];
	data?: T[];
	dataNames?: CascaderDataNames;
	onLoadData?: (options: T[]) => void;
	onConfirm?: (value: CascaderValue[], options: T[]) => void;
	onCancel?: () => void;
	onChange?: (value: CascaderValue[], options: T[]) => void;
	title?: ReactNode;
	placeholder?: ReactNode;
	swipeable?: boolean;
	closeIcon?: ReactNode;
	activeColor?: string;
}

export type CascaderProps<T = CascaderOption> = CascaderComponentProps<T> &
	Omit<HTMLAttributes<HTMLDivElement>, 'defaultValue' | 'title' | 'onChange'>;
