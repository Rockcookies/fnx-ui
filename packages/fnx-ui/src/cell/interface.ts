import { HTMLAttributes, ReactNode } from 'react';
import CellGroup from './CellGroup';

export type CellItemsAlign = 'top' | 'middle' | 'bottom';

export interface CellComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	content?: ReactNode;
	contentProps?: HTMLAttributes<HTMLDivElement>;
	description?: ReactNode;
	descriptionProps?: HTMLAttributes<HTMLDivElement>;
	leftIcon?: ReactNode;
	leftIconProps?: HTMLAttributes<HTMLSpanElement>;
	rightIcon?: ReactNode;
	rightIconProps?: HTMLAttributes<HTMLSpanElement>;
	border?: boolean;
	itemsAlign?: CellItemsAlign;
	clickable?: boolean;
	size?: 'md' | 'lg';
}

export type CellProps = CellComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellGroupComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	border?: boolean;
	insert?: boolean;
	bodyProps?: HTMLAttributes<HTMLDivElement>;
}

export type CellGroupProps = CellGroupComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellExportExtra {
	Group: typeof CellGroup;
}
