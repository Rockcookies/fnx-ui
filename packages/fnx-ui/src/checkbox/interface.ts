import { HTMLAttributes, ReactNode } from 'react';
import InternalCheckboxGroup from './CheckboxGroup';

export type CheckboxValue = string | number | boolean;

export interface CheckboxBaseProps {
	disabled?: boolean;
	direction?: 'vertical' | 'horizontal';
	checkedIcon?: ReactNode;
	icon?: ReactNode;
	iconSize?: string | number;
	iconPosition?: 'left' | 'right';
	iconShape?: 'square' | 'round' | 'plain';
	iconCheckedColor?: string;
	labelDisabled?: boolean;
}

export interface CheckboxComponentProps extends CheckboxBaseProps {
	value?: CheckboxValue;
	checked?: boolean;
	defaultChecked?: boolean;
	onChange?: (checked: boolean) => void;
	skipGroup?: boolean;
}

export type CheckboxProps = CheckboxComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'value'>;

export interface CheckboxRef {
	element: HTMLDivElement | null;
	toggle: () => void;
}

export interface CheckboxGroupComponentProps<
	T extends CheckboxValue = CheckboxValue,
> extends CheckboxBaseProps {
	value?: T[];
	defaultValue?: T[];
	onChange?: (v: T[]) => void;
	maxCheckedCount?: number;
}

export type CheckboxGroupProps<T extends CheckboxValue = CheckboxValue> =
	CheckboxGroupComponentProps<T> &
		Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'defaultValue'>;

export interface CheckboxGroupContextData extends CheckboxBaseProps {
	value?: CheckboxValue[];
	registerValue?: (value: CheckboxValue) => void;
	cancelValue?: (value: CheckboxValue) => void;
	toggleValue?: (isCheck: boolean, value: CheckboxValue) => void;
}
export interface CheckboxGroupRef {
	element: HTMLDivElement | null;
	toggleAll: (options?: { checkAll?: boolean }) => void;
}

export interface CheckboxExportExtra {
	Group: typeof InternalCheckboxGroup;
}
