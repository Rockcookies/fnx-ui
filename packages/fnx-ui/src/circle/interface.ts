import { HTMLAttributes } from 'react';

export type CircleLinecap = 'butt' | 'round' | 'square';

export type CircleGapPosition = 'top' | 'right' | 'bottom' | 'left';

export type CircleStringGradients = { [percentage: string]: string };

export interface CircleComponentProps {
	size?: string | number;
	strokeColor?: string | CircleStringGradients;
	strokeLinecap?: CircleLinecap;
	strokeWidth?: string | number;
	trailColor?: string;
	progress?: number;
	clockwise?: boolean;
	speed?: number;
	fill?: string;
	gapDegree?: number;
	gapPosition?: CircleGapPosition;
}

export type CircleProps = CircleComponentProps & HTMLAttributes<HTMLDivElement>;
