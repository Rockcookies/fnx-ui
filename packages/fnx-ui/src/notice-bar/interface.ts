import { HTMLAttributes, ReactNode } from 'react';

export type NoticeBarProps = NoticeBarComponentProps &
	HTMLAttributes<HTMLDivElement>;

export interface NoticeBarComponentProps {
	marquee?: boolean | 'auto';
	marqueeDelay?: number;
	marqueeSpeed?: number;
	ellipsis?: boolean;
	leftIcon?: ReactNode;
	rightIcon?: ReactNode;
	closeable?: boolean;
	closeIcon?: ReactNode;
	color?: string;
	background?: string;
}

export interface NoticeBarRef {
	element: HTMLDivElement | null;
	reset: () => void;
}
