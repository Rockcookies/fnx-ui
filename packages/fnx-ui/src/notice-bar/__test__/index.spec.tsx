import { act, render, fireEvent, waitFor } from '@testing-library/react';
import React from 'react';
import NoticeBar from '../index';
import { NoticeBarRef } from '../interface';
import TestsReact from '../../../test/react';
import TestsDOM from '../../../test/dom';
describe('<NoticeBar/>', () => {
	const mockRect = (el: Element, rect: Partial<DOMRect>) => {
		Object.defineProperty(el, 'getBoundingClientRect', {
			configurable: true,
			get: () => {
				return () => ({
					top: 0,
					left: 0,
					right: 0,
					bottom: 0,
					width: 0,
					height: 0,
					...rect,
				});
			},
		});
	};

	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterAll(() => {
		jest.useRealTimers();
	});

	it('clear noticeBar element when close icon is clicked', () => {
		const { container } = render(
			<NoticeBar marquee closeable>
				Content
			</NoticeBar>,
		);

		const target = TestsDOM.mustQuerySelector(
			container,
			'.fnx-notice-bar__close-icon',
		);

		fireEvent.click(target);

		expect(container).toMatchSnapshot();
	});

	it('should render children correct', () => {
		const { container } = render(
			<NoticeBar marqueeDelay={0} marquee>
				Content
			</NoticeBar>,
		);

		expect(container).toMatchSnapshot();
	});

	it('should not start scrolling when marquee false', () => {
		const { container } = render(
			<NoticeBar marqueeDelay={0} marquee={false}>
				Content
			</NoticeBar>,
		);

		expect(container).toMatchSnapshot();
	});

	it('should start scrolling when marquee auto and marquee width > marquee-wrapper width', () => {
		const { container } = render(
			<NoticeBar marquee="auto" marqueeDelay={0}>
				Content
			</NoticeBar>,
		);

		const wrapper = TestsDOM.mustQuerySelector(
			container,
			'.fnx-notice-bar__marquee-wrapper',
		);

		const content = TestsDOM.mustQuerySelector(
			container,
			'.fnx-notice-bar__marquee',
		);

		mockRect(wrapper, {
			width: 50,
			height: 50,
		});

		mockRect(content, {
			width: 100,
			height: 50,
		});

		act(() => {
			jest.runTimersToTime(2000);
		});

		expect(container).toMatchSnapshot();
	});

	it('should expose reset methods', () => {
		const NoticeBarRef = React.createRef<NoticeBarRef>();

		const {} = render(
			<NoticeBar ref={NoticeBarRef} marqueeDelay={0} marquee>
				Content
			</NoticeBar>,
		);

		act(() => {
			TestsReact.mustGetRef(NoticeBarRef).reset();
		});

		expect(TestsReact.mustGetRef(NoticeBarRef).reset).toBeTruthy();
	});
});
