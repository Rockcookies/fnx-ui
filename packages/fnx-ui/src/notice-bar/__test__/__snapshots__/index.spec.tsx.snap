// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoticeBar/> clear noticeBar element when close icon is clicked 1`] = `<div />`;

exports[`<NoticeBar/> should not start scrolling when marquee false 1`] = `
<div>
  <div
    class="fnx-notice-bar"
  >
    <div
      class="fnx-notice-bar__content-wrapper"
    >
      <div
        class="fnx-notice-bar__content fnx-notice-bar__content--wrap-text"
      >
        Content
      </div>
    </div>
  </div>
</div>
`;

exports[`<NoticeBar/> should render children correct 1`] = `
<div>
  <div
    class="fnx-notice-bar fnx-notice-bar--marquee"
  >
    <div
      class="fnx-notice-bar__marquee-wrapper"
    >
      <div
        class="fnx-notice-bar__marquee"
      >
        Content
      </div>
    </div>
  </div>
</div>
`;

exports[`<NoticeBar/> should start scrolling when marquee auto and marquee width > marquee-wrapper width 1`] = `
<div>
  <div
    class="fnx-notice-bar fnx-notice-bar--marquee"
  >
    <div
      class="fnx-notice-bar__marquee-wrapper"
    >
      <div
        class="fnx-notice-bar__marquee"
        style="transform: translateX(-100px); transition-duration: 1666.6666666666667ms;"
      >
        Content
      </div>
    </div>
  </div>
</div>
`;
